crates/ecs/src/filter.rs

```diff
@@ -0,0 +1,345 @@
+    type BorrowedData<'components> = ReadComponentVec<'components, Component>;
+
+    fn borrow(world: &World) -> Self::BorrowedData<'_> {
+        world.borrow_component_vec::<Component>()
+    }
+
+    unsafe fn fetch_parameter(
+        borrowed: &mut Self::BorrowedData<'_>,
+        entity: Entity,
+    ) -> Option<Self> {
+        if let Some(component_vec) = borrowed {
+            if let Some(Some(_)) = component_vec.get(entity.id) {
+                return Some(Self {
+                    phantom: Default::default(),
+                });
+            }
+        }
+        None
+    }
+
+    fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+        vec![ComponentAccessDescriptor::Read(TypeId::of::<Component>())]
+    }
+}
+
+/// A query filter that matches any entity without a given component type.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::{With, Without};
+/// # use ecs::Read;
+/// # #[derive(Debug)]
+/// # struct Position;
+/// # struct Player;
+/// fn non_player_position(position: Read<Position>, _: Without<Player>) {
+///     println!("A non-player entity is at position {:?}.", position);
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Without<Component: 'static> {
+    phantom: PhantomData<Component>,
+}
+
+impl<Component: Send + Sync + 'static + Sized> SystemParameter for Without<Component> {
+    type BorrowedData<'components> = ReadComponentVec<'components, Component>;
+
+    fn borrow(world: &World) -> Self::BorrowedData<'_> {
+        world.borrow_component_vec::<Component>()
+    }
+
+    unsafe fn fetch_parameter(
+        borrowed: &mut Self::BorrowedData<'_>,
+        entity: Entity,
+    ) -> Option<Self> {
+        if let Some(component_vec) = borrowed {
+            if let Some(Some(_)) = component_vec.get(entity.id) {
+                return None;
+            }
+        }
+        Some(Self {
+            phantom: Default::default(),
+        })
+    }
+
+    fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+        vec![ComponentAccessDescriptor::Read(TypeId::of::<Component>())]
+    }
+}
+
+macro_rules! binary_filter_operation {
+    ($name:ident, $op:tt, $op_name:literal, $op_name_lowercase:literal) => {
+        /// A query filter that combines two filters using the `
+		#[doc = $op_name]
+		///` operation.
+        ///
+        #[doc = concat!(
+            "# Example\n```\n",
+            "# use {ecs::filter::With, ecs::filter::", stringify!($name), "};\n",
+            "# use ecs::Read;\n",
+            "# #[derive(Debug)]\n",
+            "# struct Position;\n",
+            "# struct Player;\n",
+            "# struct Enemy;\n",
+            "fn player_", $op_name_lowercase, "_enemy(position: Read<Position>, _: ", stringify!($name), "<With<Player>, With<Enemy>>) {\n",
+            "    println!(\"An entity at position {:?} is a player ", $op_name_lowercase ," an enemy.\", position);\n",
+            "}\n```",
+        )]
```

Comment:
Yes, doc-comments generated from macros are visible, but there are currently some limitations within IntelliJ-Rust when using `concat!` and `stringify!`, so the example will not show up in CLion, but it will be visible in the generated documentation from `cargo doc`.

Reply:
![image](https://user-images.githubusercontent.com/12572888/226582214-8c5607c0-76e5-4575-9d9d-b902fc772f45.png)
Is what I'm hearing... 😝 

Alright, well that's better than no documentation at least! 👍